README.md

[![Arduino CI](https://github.com/RobTillaart/PCF8591/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/PCF8591/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/PCF8591/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/PCF8591/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/PCF8591/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/PCF8591/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/PCF8591.svg?maxAge=3600)](https://github.com/RobTillaart/PCF8591/releases)


# PCF8591

Arduino Library for PCF8591 I2C 4 channel 8 bit ADC + 1 channel 8 bit DAC.


## Description

**warning** during tests I could overclock the PCF8591 chip up to 650 KHz.
However it is only specified to run at 100 kHz. 
After some time it was getting pretty hot and it broke down. 
So overclocking is fun but not recommended.

PCF8591 has one 8 bit ADC on board for 4 channels. The ADC is 8 bit and quite fast.
At 100 KHz one gets \> 2000 reads per second for **analogRead()** and 
\> 2000 writes per second for **analogWrite()**.  
Note that most time is probably spend on I2C communication.

The auto-increment functionality is used in the **analogRead4()** function.
The library only supports it for the mode 0 (plain ADC, no differential). 
The **lastRead()** function is needed to get access to the values.
First tests shows it is 2.6 x faster than 4 individual reads.


## Interface

```cpp
#include "PCF8591.h"
```


#### Constructor

- **PCF8591(const uint8_t address = 0x48, TwoWire \*wire = &Wire)** constructor with I2C address.
Default is 0x48, optional set the WireN I2C bus.
- **bool begin(uint8_t sda, uint8_t scl, uint8_t value = 0)** set wire pins for ESP series.
Also set initial value for the DAC. 
Returns **true** if successful.
- **bool begin(uint8_t value = 0)** Set initial value for the DAC. 
Returns **true** if successful.
- **bool isConnected()** test to see if chip can be reached.


#### ADC channels

The PCF8591 has four 8 bit ADC channels. Values = 0..255.

- **void enableINCR()** used in analogRead4(). 
Could become private in the future.
- **void disableINCR()** idem.
- **bool isINCREnabled()** idem.
- **uint8_t analogRead(uint8_t channel, uint8_t mode = 0)** read one of the 4 analogue ports.  
Default mode is PCF8591_FOUR_SINGLE_CHANNEL, see table below.
For details comparator modes see datasheet figure 4 page 6.
- **uint8_t analogRead4()** read all 4 channels in one call.
Works in PCF8591_FOUR_SINGLE_CHANNEL mode only.
Uses **enableINCR()** to do that efficiently. 
It is about 2.6 times faster than 4 individual **analogRead()** calls.
The latter allows for optimized timing per channel and the order 
in which the channels are read.
Use **lastRead()** to access the 4 values.
Returns **PCF8591_OK** or an error code if appropriate.
- **uint8_t lastRead(uint8_t channel)** get last read value from cache.  
This cache is filled both by **analogRead()** and **analogRead4()**. 
See example sketch.


|  ADC mode                     |  Value  |  Notes    |
|:------------------------------|:-------:|:----------|
|  PCF8591_FOUR_SINGLE_CHANNEL  |  0x00   |  default  |
|  PCF8591_THREE_DIFFERENTIAL   |  0x01   |
|  PCF8591_MIXED                |  0x02   |
|  PCF8591_TWO_DIFFERENTIAL     |  0x03   |


#### Comparator

Since 0.2.0 four direct comparator calls are added to make the code more explicit.
These four calls are in fact wrappers around the **analogRead()**.
These still need to be tested as I had no hardware available.

- **int readComparator_01()** == analogRead(channel = 0, mode = 3);
- **int readComparator_23()** == analogRead(channel = 1, mode = 3);
- **int readComparator_03()** == analogRead(channel = 0, mode = 1);
- **int readComparator_13()** == analogRead(channel = 1, mode = 1);

The values of the comparator calls are cached and can be accessed with **lastRead()**.
Be sure to select the right channel for **lastRead()**.

Note: these functions do never use PCF8591_MIXED (2) mode.


#### DAC channel

The PCF8591 has one 8 bit DAC. output value 0..255 == 0..Vref Volts (datasheet).

- **void enableDAC()** switch on the analogue output.
- **void disableDAC()** switch off the analogue output (high impedance). Sort of energy saving mode.
- **bool isDACEnabled()** check the modus operandi.
- **bool analogWrite(uint8_t value = 0)** writes a value 0..255 to the DAC. Check datasheet for voltage.
Note, this is a real voltage not a PWM signal like **analogWrite()** on an UNO.
- **uint8_t lastWrite()** get last value written (from cache).


#### Error codes

- **int lastError()** always check this value after a read / write to see if it was OK (== 0).
After the read the error value is reset to OK.

To elaborate

|  Error code             |  Value  |  Notes  |
|:------------------------|:-------:|:--------|
|  PCF8591_OK             |  0x00   |
|  PCF8591_PIN_ERROR      |  0x81   |
|  PCF8591_I2C_ERROR      |  0x82   |
|  PCF8591_MODE_ERROR     |  0x83   |
|  PCF8591_CHANNEL_ERROR  |  0x84   |
|  PCF8591_ADDRESS_ERROR  |  0x85   |


## Operations

See examples.


## Future

#### Must

- improve documentation
- test / verify comparator calls with hardware.
  - datasheet (par.8.2 figure 4)

#### Should

- add examples for comparator calls.
  - schema?


#### Could


#### Wont




CHANGELOG.md
# Change Log PCF8591

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.0] - 2023-02-23
- fix#10 analogRead() for differential modi.
- add defines for 4 different modi
- add four comparator functions (wrappers).
- update readme.md
- update GitHub actions
- update license 2023
- minor edits

----

## [0.1.3] - 2022-11-21
- add RP2040 to build-CI
- add changelog.md
- move code from .h to .cpp
- update unit test

## [0.1.2] - 2021-12-23
- update library.json
- update readme.md
- update license
- minor edits

## [0.1.1] - 2021-01-14
- added WireN 
- improve error handling.

## [0.1.0] - 2021-01-04
- add Arduino-CI

----

## [0.0.2] - 2020-07-22
- testing
- refactor
- add documentation and examples

## [0.0.1] - 2020-03-12
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2021-01-01
// PURPOSE: unit tests for the I2C PCF8591 library
//          https://github.com/RobTillaart/PCF8591
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "PCF8591.h"



unittest_setup()
{
}

unittest_teardown()
{
}


unittest(test_constructor)
{
  fprintf(stderr, "VERSION: %s\n", PCF8591_LIB_VERSION);
  
  PCF8591 dev(0x48);

  assertTrue(dev.begin());
  assertTrue(dev.isConnected());
  assertEqual(0, dev.lastWrite());
  assertEqual(PCF8591_OK, dev.lastError());
}


unittest(test_ADC_INCR)
{
  fprintf(stderr, "VERSION: %s\n", PCF8591_LIB_VERSION);
  
  PCF8591 dev(0x48);

  assertTrue(dev.isINCREnabled());
  dev.enableINCR();
  assertTrue(dev.isINCREnabled());
  dev.disableINCR();
  assertFalse(dev.isINCREnabled());
}


unittest(test_DAC)
{
  fprintf(stderr, "VERSION: %s\n", PCF8591_LIB_VERSION);
  
  PCF8591 dev(0x48);

  assertTrue(dev.isDACEnabled());
  dev.enableDAC();
  assertTrue(dev.isDACEnabled());
  dev.disableDAC();
  assertFalse(dev.isDACEnabled());
}


unittest_main()

// --------


PCF8591.cpp
//
//    FILE: PCF8591.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-03-12
// VERSION: 0.2.0
// PURPOSE: I2C PCF8591 library for Arduino
//     URL: https://github.com/RobTillaart/PCF8591


#include "PCF8591.h"


//  INTERNAL USE ONLY
#define PCF8591_DAC_FLAG                0x40
#define PCF8591_INCR_FLAG               0x04


PCF8591::PCF8591(const uint8_t address, TwoWire *wire)
{
  if ((address < 0x48) || (address > 0x4F))
  {
    _error = PCF8591_ADDRESS_ERROR;
    return;
  }
  _address = address;
  _wire    = wire;
  _control = 0;
  _dac     = 0;
  for (uint8_t i = 0; i < 4; i++)
  {
    _adc[i] = 0;
  }
  _error = PCF8591_OK;
}


#if defined (ESP8266) || defined(ESP32)
bool PCF8591::begin(uint8_t sda, uint8_t scl, uint8_t val)
{
  _wire = &Wire;
  if ((sda < 255) && (scl < 255))
  {
    _wire->begin(sda, scl);
  } else {
    _wire->begin();
  }
  if (!isConnected()) return false;
  analogWrite(val);
  return true;
}
#endif


bool PCF8591::begin(uint8_t val)
{
  _wire->begin();
  if (!isConnected()) return false;
  analogWrite(val);
  return true;
}


bool PCF8591::isConnected()
{
  _wire->beginTransmission(_address);
  _error = _wire->endTransmission();  //  default == 0 == PCF8591_OK
  return( _error == PCF8591_OK);
}


//////////////////////////////////////////////////////////
//
//  ADC PART
//
void PCF8591::enableINCR()
{
  _control |= PCF8591_INCR_FLAG;
}


void PCF8591::disableINCR()
{
  _control &= ~PCF8591_INCR_FLAG;
}


bool PCF8591::isINCREnabled()
{
  return ((_control & PCF8591_INCR_FLAG) > 0);
}


uint8_t PCF8591::analogRead(uint8_t channel, uint8_t mode)
{
  if (mode > 3)
  {
    _error = PCF8591_MODE_ERROR;
    return 0;
  }
  _control &= 0b01000100;         //  clear all except flags
  _control |= (mode << 4);

  _error = PCF8591_CHANNEL_ERROR;
  switch(mode)
  {
    case PCF8591_FOUR_SINGLE_CHANNEL:
      if (channel > 3) return 0;
      break;
    case PCF8591_THREE_DIFFERENTIAL:
      if (channel > 2) return 0;
      break;
    case PCF8591_MIXED:
      if (channel > 2) return 0;
      break;
    case PCF8591_TWO_DIFFERENTIAL:
      if (channel > 1) return 0;
      break;
    default:
      return 0;
  }
  _control |= channel;
  _error = PCF8591_OK;

  //  NOTE: one must read two values to get an up to date value.
  //        Page 8 datasheet.
  _wire->beginTransmission(_address);
  _wire->write(_control);
  _error = _wire->endTransmission();  // default == 0 == PCF8591_OK
  if (_error != 0) return PCF8591_I2C_ERROR;

  if (_wire->requestFrom(_address, (uint8_t)2) != 2)
  {
    _error = PCF8591_I2C_ERROR;
    return _adc[channel];          // known last value
  }
  _wire->read();
  _adc[channel] = _wire->read();
  return _adc[channel];
}


uint8_t PCF8591::analogRead4()
{
  //  clear all except flags
  //  MODE == PCF8591_FOUR_SINGLE_CHANNEL.
  _control &= 0b01000100;
  uint8_t channel = 0;
  _control |= channel;

  enableINCR();
  _wire->beginTransmission(_address);
  _wire->write(_control);
  _error = _wire->endTransmission();  // default == 0 == PCF8591_OK
  if (_error != 0)
  {
    _error = PCF8591_I2C_ERROR;
    disableINCR();
    return _error;
  }
  if (_wire->requestFrom(_address, (uint8_t)5) != 5)
  {
    _error = PCF8591_I2C_ERROR;
    disableINCR();
    return _error;
  }

  _wire->read();
  for (uint8_t i = 0; i < 4; i++)
  {
    _adc[i] = _wire->read();
  }
  _error = PCF8591_OK;
  disableINCR();
  return _error;
}


uint8_t PCF8591::lastRead(uint8_t channel)
{
  return _adc[channel];
};


//  comparator calls need testing.
int PCF8591::readComparator_01()
{
  int8_t v = analogRead(0, 3);
  return v;
}


int PCF8591::readComparator_23()
{
  int8_t v = analogRead(1, 3);
  return v;
}


int PCF8591::readComparator_03()
{
  int8_t v = analogRead(0, 1);
  return v;
}


int PCF8591::readComparator_13()
{
  int8_t v = analogRead(1, 1);
  return v;
}



//////////////////////////////////////////////////////////
//
//  DAC PART
//
void PCF8591::enableDAC()
{
  _control |= PCF8591_DAC_FLAG;
};


void PCF8591::disableDAC()
{
  _control &= ~PCF8591_DAC_FLAG;
};


bool PCF8591::isDACEnabled()
{
  return ((_control & PCF8591_DAC_FLAG) > 0);
};


bool PCF8591::analogWrite(uint8_t value)
{
  _wire->beginTransmission(_address);
  _wire->write(_control);
  _wire->write(value);
  _error = _wire->endTransmission();
  if (_error != 0)
  {
    _error = PCF8591_I2C_ERROR;
    return false;
  }
  _dac = value;
  return true;
}


uint8_t PCF8591::lastWrite()
{
  return _dac;
};


//////////////////////////////////////////////////////////
//
//  ERROR HANDLING
//
int PCF8591::lastError()
{
  int e = _error;
  _error = PCF8591_OK;
  return e;
}


//  -- END OF FILE --



PCF8591.h
#pragma once
//
//    FILE: PCF8591.h
//  AUTHOR: Rob Tillaart
//    DATE: 2020-03-12
// VERSION: 0.2.0
// PURPOSE: I2C PCF8591 library for Arduino
//     URL: https://github.com/RobTillaart/PCF8591


#include "Arduino.h"
#include "Wire.h"


#define PCF8591_LIB_VERSION                 (F("0.2.0"))

#define PCF8591_OK                          0x00
#define PCF8591_PIN_ERROR                   0x81
#define PCF8591_I2C_ERROR                   0x82
#define PCF8591_MODE_ERROR                  0x83
#define PCF8591_CHANNEL_ERROR               0x84
#define PCF8591_ADDRESS_ERROR               0x85

//  datasheet figure 4 page 6
#define PCF8591_FOUR_SINGLE_CHANNEL         0x00  //  default
#define PCF8591_THREE_DIFFERENTIAL          0x01
#define PCF8591_MIXED                       0x02
#define PCF8591_TWO_DIFFERENTIAL            0x03


class PCF8591
{
public:
  explicit PCF8591(const uint8_t address = 0x48, TwoWire *wire = &Wire);

#if defined (ESP8266) || defined(ESP32)
  bool     begin(uint8_t sda, uint8_t scl, uint8_t value = 0);
#endif
  bool     begin(uint8_t value = 0);

  bool     isConnected();


  //       ADC PART
  //       auto increment not tested ==> use with care!
  void     enableINCR();
  void     disableINCR();
  bool     isINCREnabled();

  //       analogRead() returns the value.
  //       mode 0 = PCF8591_FOUR_SINGLE_CHANNEL
  uint8_t  analogRead(uint8_t channel, uint8_t mode = 0);
  //       analogRead4() returns PCF8591_OK or an error code.
  uint8_t  analogRead4();
  //       access the 4 channels read with analogRead4()
  uint8_t  lastRead(uint8_t channel);

  //       datasheet par 8.2 figure 4
  //       not for PCF8591_MIXED mode.
  //       comparator calls need testing.
  int      readComparator_01();  //  channel 0  mode 3
  int      readComparator_23();  //  channel 1  mode 3
  int      readComparator_03();  //  channel 0  mode 1
  int      readComparator_13();  //  channel 1  mode 1


  //       DAC PART
  void     enableDAC();
  void     disableDAC();
  bool     isDACEnabled();

  bool     analogWrite(uint8_t value = 0);  //  returns true on success.
  uint8_t  lastWrite();                     //  returns last successful write

  //       ERROR HANDLING
  int      lastError();


private:
  uint8_t  _address;
  uint8_t  _control;
  uint8_t  _dac;
  uint8_t  _adc[4];
  int      _error;

  TwoWire* _wire;
};


//  -- END OF FILE --

